model/task_queue_service_dependency.go

```diff
@@ -0,0 +1,314 @@
+		})
+	}()
+	t.lastUpdated = time.Now()
+
+	// Add each individual node (TaskQueueItem) to the graph
+	for i := range items {
+		t.addItem(&items[i])
+	}
+
+	// Save task groups
+	t.taskGroups = map[string]taskGroupTasks{}
+	for _, item := range items {
+		if item.Group != "" {
+			id := compositeGroupId(item.Group, item.BuildVariant, item.Version)
+			if _, ok := t.taskGroups[id]; !ok {
+				t.taskGroups[id] = taskGroupTasks{
+					id:       id,
+					maxHosts: item.GroupMaxHosts,
+					tasks:    []TaskQueueItem{item},
+				}
+			} else {
+				taskGroup := t.taskGroups[id]
+				taskGroup.tasks = append(taskGroup.tasks, item)
+				t.taskGroups[id] = taskGroup
+			}
+		}
+	}
+
+	// Add edges for task dependencies
+	for _, item := range items {
+		for _, dep := range item.Dependencies {
+			t.addEdge(item.Id, dep)
+		}
+	}
+
+	// Sort the graph. Use a lexical sort to resolve ambiguities, because node order is the
+	// order that we received these in.
+	sorted, err := topo.SortStabilized(t.graph, lexicalReversed)
+	if err != nil {
+		grip.Alert(message.WrapError(err, message.Fields{
+			"message": "problem sorting tasks",
+		}))
+		return
+	}
+	t.sorted = sorted
+
+	return
+}
+
+type byIDReversed []graph.Node
+
+func (n byIDReversed) Len() int           { return len(n) }
+func (n byIDReversed) Less(i, j int) bool { return n[i].ID() > n[j].ID() }
+func (n byIDReversed) Swap(i, j int)      { n[i], n[j] = n[j], n[i] }
+func lexicalReversed(nodes []graph.Node)  { sort.Sort(byIDReversed(nodes)) }
+
+// FindNextTask returns the next dispatchable task in the queue.
+func (t *taskDistroDAGDispatchService) FindNextTask(spec TaskSpec) *TaskQueueItem {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// If the host just ran a task group, give it one back
+	if spec.Group != "" {
+		taskGroup, ok := t.taskGroups[compositeGroupId(spec.Group, spec.BuildVariant, spec.Version)]
+		if ok {
+			if next := t.nextTaskGroupTask(taskGroup); next != nil {
+				node := t.getNodeByItemID(next.Id)
+				item := t.getItemByNodeID(node.ID())
+				item.IsDispatched = true
+				return next
+			}
+		}
+		// If the task group is not present in the task group map, it has been dispatched,
+		// so fall through to getting a task not in that group.
+	}
+
+	// Iterate backwards through the graph, because dependencies are on the left.
+	dependencyCaches := make(map[string]task.Task)
+	for i := len(t.sorted) - 1; i >= 0; i-- {
+		node := t.sorted[i]
+		item := t.getItemByNodeID(node.ID())
+
+		// TODO Consider checking if the state of any task has changed, which could unblock
+		// later tasks in the queue. Currently we just wait for the scheduler to rerun.
+
+		// For a task not in a task group, dispatch it if it hasn't been dispatched, and if
+		// its dependencies have been met
+		if item.GroupMaxHosts == 0 {
+			if item.IsDispatched {
+				continue
+			}
+			nextTaskFromDB, err := task.FindOneId(item.Id)
+			depsMet, err := nextTaskFromDB.DependenciesMet(dependencyCaches)
```

Comment:
@StuartButler I notice that we check `Dependencies` met here, but not for task group tasks where maxHosts > 1. Do you think checking dependencies makes sense for that case?

Reply:
As discussed in person: for this case, we will leave it as is for now to be revisited soon to add this functionality.